docs/design_proposals/sync-auto-support.md

```diff
@@ -0,0 +1,255 @@
+
+3. Is there any another workaround, and if so, what are its drawbacks?
+  - Currently one can create a docker file that only copies specific build
+    results into the container and relies on a local build to generate those
+    intermediate artifacts. This requires a user to trigger a first build
+    manually before docker kicks in to containerize the application. While
+    it may be possible to automate this (for example: gradle --continuous), it
+    is not an acceptable solution to require manual external processes for
+    a build to succeed. Covered in [Hack it](#hack-it)
+
+4. Mention related issues, if there are any.
+  - This is not trying to solve the problem of dealing with a multistage
+    dockerbuild. Intermediate build artifacts might still be possible to
+    determine, however that would require an extra mechanism to do a partial
+    docker build and sync files from a built container to a running container --
+    something we do not intend to cover here.
+
+#### Problems with current API/config
+The current `sync` system has the following problems:
+1. No way to trigger local external processes - for example, in jib, to build a
+   container, one would run `./gradlew jib`, but to update class files so the
+   system may sync them, one would only be required to run `./gradlew classes`.
+2. No way to tell the system to sync non build inputs. Skaffold is
+   normally only watching `.java` files for a build, in the sync case, we want
+   it to watch `.java` files, trigger a partial build, and sync `.class` files
+   so a remote server can pick up and reload the changes.
+
+## Design
+
+### Hack it
+To get close to the functionality we want, without modifying skaffold at all, a
+Dockerfile which depends on java build outputs could be used, like:
+```
+FROM openjdk:8
+COPY build/dependencies/ /app/dependencies
+COPY build/classes/java/main/ /app/classes
+
+CMD ["java", "-cp", "/app/classes:/app/dependencies/*", "hello.Application"]
+```
+
+with a skaffold sync block that looks like:
+```
+sync:
+  manual:
+    - src: "build/classes/java/main/**/*.class"
+      dest: "/app/classes"
+      strip: "build/classes/java/main/"
+```
+
+A user's devloop then looks like this:
+
+1. run `./gradlew classes copyDependencies`
+1. run `skaffold dev`
+1. *make changes to some java file*
+1. run `./gradlew classes`
+1. *skaffold syncs files*
+
+which is far from ideal.
+
+### Fix it?
+
+Provide users with an `auto` option, users that use `auto` should expect
+the builder-sync to do the right thing and will not be required to do much
+configuration.
+
+Auto will only work with builders that have implemented the auto spec. We
+expect at least `jib` to do implement the spec. How `auto` is implemented can be
+in the following ways.
+
+#### User Configuration
+
+```yaml
+build:
+  artifacts:
+  - image: ...
+    context: jib-project
+    jib: {}
+    sync:
+      auto: {}
+```
+
+#### Option 1: Delegate generation of sync tar to builder
```

Comment:
@corneliusweig @tejal29 added in this section about the builder being responsible for the syncable tar...

I can see the problem that it kind of ties down the sync implementation to tar.